cluster/assemblestate/assemblestate.go

```diff
@@ -0,0 +1,725 @@
+		as.selector.AddAuthoritativeRoute(auth.RDT, addr)
+	}
+
+	as.commit()
+
+	as.logger.Debug("got valid auth message", "peer-rdt", auth.RDT)
+
+	return nil
+}
+
+// VerifyPeer checks if the given certificate is trusted and maps to a known RDT.
+// If it is, then a [PeerHandle] is returned that can be used to modify the state
+// of the cluster on this peer's behalf.
+//
+// An error is returned if the certificate isn't trusted.
+func (as *AssembleState) VerifyPeer(cert []byte) (*PeerHandle, error) {
+	as.lock.Lock()
+	defer as.lock.Unlock()
+
+	fp := CalculateFP(cert)
+
+	p, ok := as.trusted[fp]
+	if !ok {
+		return nil, errors.New("given TLS certificate is not associated with a trusted RDT")
+	}
+
+	return &PeerHandle{
+		as:   as,
+		peer: p.RDT,
+	}, nil
+}
+
+// Run starts the assembly process, managing both the server and periodic client operations.
+// It returns when the context is cancelled, returning the final routes discovered.
+func (as *AssembleState) Run(
+	ctx context.Context,
+	transport Transport,
+	discover Discoverer,
+) (Routes, error) {
+	addr := fmt.Sprintf("%s:%d", as.config.IP, as.config.Port)
+	client := transport.NewClient(as.cert)
+
+	var wg sync.WaitGroup
+
+	// start the server that handles incoming requests
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if err := transport.Serve(ctx, addr, as.cert, as); err != nil {
+			as.logger.Error(err.Error())
+		}
+	}()
+
+	// start periodic discovery of peers.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		periodic(ctx, time.Second*5, time.Second*1, func(ctx context.Context) {
+			discoveries, err := discover(ctx)
+			if err != nil {
+				as.logger.Error(err.Error())
+				return
+			}
+
+			// filter out our address
+			addrs := make([]string, 0, len(discoveries))
+			for _, d := range discoveries {
+				if d == addr {
+					continue
+				}
+				addrs = append(addrs, d)
+			}
+
+			if err := as.publishAuth(ctx, addrs, client); err != nil {
+				as.logger.Error(err.Error())
+				return
+			}
+		})
+	}()
+
+	// start up the periodic publication of routes
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		const (
+			period = time.Second * 5
+			jitter = time.Second
+			peers  = 5
+			routes = 5000
+		)
+		periodic(ctx, period, jitter, func(ctx context.Context) {
+			as.publishRoutes(ctx, client, peers, routes)
+		})
+	}()
+
+	// start event-driven device operations
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		for {
+			select {
+			case <-as.devices.Responses():
+				as.publishDevices(ctx, client)
+			case <-as.devices.Queries():
+				as.publishDeviceQueries(ctx, client)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// wait for context cancellation
+	wg.Wait()
+
+	sent, received := transport.Stats()
+	as.logger.Info("assemble stopped",
+		"sent-bytes", sent,
+		"received-bytes", received,
+	)
+
+	return as.selector.Routes(), nil
+}
+
+func periodic(
+	ctx context.Context,
+	interval time.Duration,
+	jitter time.Duration,
+	work func(ctx context.Context),
+) {
```

Comment:
is this equivalent or not to the Trickle logic, where you have  a period, and within it you have a listen only first half and then a point in the 2nd half that is transmission time if there's anything to transmit?